Create a screenbuf and a new shell

xlmod opened the pull request:
Create a big buffer with scroll and multi screen capacity.
And create a new shell with arguments and many builtins.

Closes #39

xlmod committed:
Create Screen Buffer with scroll

New screen buffer with multiple buffer possibility and scroll.
Add function in vga to print a char in color with a offset.
Add define in vga.h with pointer of the vga buffer and size of the size
of the buffer.
Use the new buffer in the kprintf function.

Signed-off-by: xlmod <glafond-@student.42.fr>

mrxx0 committed:
Fix scrolling on screen buffer

Signed-off-by: mrxx0 <chcoutur@student.42.fr>

xlmod committed:
Setup a nice screenbuf interface and intergrate it

Change print and kernel_main to use screenbuf insteed of the vga lib.

Signed-off-by: xlmod <glafond-@student.42.fr>

xlmod committed:
Add cursor handling and '\t' and '\b'

Update cursor when write charactere.
Enable or disable cursor when cursor is on the screen or not.
'\t': add between 1-4 space to put cursor on mod 4 value.
'\b': remove th last charactere, if the start of the line put the cursor
at the last charactere of the previous line.

Signed-off-by: xlmod <glafond-@student.42.fr>

xlmod committed:
Change the shell to a better one and fix strcmp.

Create ksh a simple shell with multiple arguments.
The shell call function found in an array call builtin in ksh.c and look
for a function with a name corresponmding to the first argument of the
command.
To execute a command, the shell translate the string entered by the user
to a array of string and call the correct function with the array of
string in argument of the function.
The shell is not perfectly working, all interaction with the screenbuf
execpt printing charactere doesn't work.

Fix strcmp: change `return s1 - s2;` to `return *s1 - *s2`.

Signed-off-by: glafond- <glafond-@student.42.fr>

xlmod committed:
Fix screenbuf cmd, rename nsh and split builtins

Fix screenbuf cmd: extern screenbuf struct change from * to [].
Tree restructuration of the new shell:
- remane ksh in nsh.
- create a nsh dir
- put builtins in kernel/nsh/builtins.
- change builtins prototypes to int (*f)(int, char **).

Signed-off-by: xlmod <glafond-@student.42.fr>

xlmod committed:
Add clear, next, prev and help builtins

Signed-off-by: xlmod <glafond-@student.42.fr>

e2r3p13 committed:
Rework of the screenbuf codebase

Fix coding style to match nulix coding style guidelines.

Fix the buf that occured when erasing characters after writing more
than a full circular buffer. We were using sb->top instead of sb->??
(I don't remember...)

Update structure and functions names to make things more clear.

Signed-off-by: lfalkau <lfalkau@student.42.fr>

dummy

e2r3p13 committed:
Rework of the screenbuf codebase

Fix coding style to match nulix coding style guidelines.

Fix the buf that occured when erasing characters after writing more
than a full circular buffer. We were using sb->top instead of sb->??
(I don't remember...)

Note that I only reviewed code up to line number 207. I plan to finish
the review tomorrow.

Signed-off-by: lfalkau <lfalkau@student.42.fr>

mrxx0 committed:
Add a global info function for modules (idt, gdt..)

- info() function to print information about some modules of the kernel
  such as the stack, the gdt, the idt, the buddy allocator and the x86
  registers.

Signed-off-by: mrxx0 <chcoutur@student.42.fr>

xlmod committed:
Add hexdump builtins

usage: hexdump ADDRESS SIZE

Signed-off-by: xlmod <glafond-@student.42.fr>

xlmod committed:
Remove vga.c and vga.h files

Remove all use of vga api in all files.
Move color in screenbuf.h and rename then SB_COLOR_color.
Add SB_DFL_FG/BG_COLOR and compute SB_DFL_COLOR from them.
Set SB_WHITESPACE as a 0x00 in SB_DFL_COLOR color.
Move VGA_{PTR,WIDTH,HEIGHT} in screenbuf.h as well.
Change hexdump color to default color and light red.

Signed-off-by: xlmod <glafond-@student.42.fr>

xlmod committed:
Implement interrupt builtin and fix warning.

Signed-off-by: glafond- <glafond-@student.42.fr>

e2r3p13 committed:
Prompt update / Move view to cursor on keypress

The view now down to the cursor position on key press to avoid
getting lost in the screenbuf.

The prompt is also made more fancy, and with color to be seen
easily in the middle of some output

Signed-off-by: lfalkau <lfalkau@student.42.fr>

e2r3p13 commented on kernel/nsh/builtins/color.c:
```diff
@@ -0,0 +1,72 @@
+		"lightbrown",
+		"white"
+	};
+	if (argc < 2) {
+		kprintf("color_change" ": Not enough arguments.\n");
+		return -1;
+	}
+	if (!strcmp(argv[1], "reset")) {
+		sb_set_color(sb + sb_index, SB_DFL_COLOR);
+	} else {
+		if (argc < 3) {
+			kprintf("color_change" ": Not enough arguments.\n");
+			return -1;
+		}
+		if (!strcmp(argv[1], "fg")) {
+			for (int i = 0; i < 16; i++) {
+				if (!strcmp(colors[i], argv[2]))
+					sb_set_fg(sb + sb_index, i);
+			}
+		} else if (!strcmp(argv[1], "bg")) {
+			for (int i = 0; i < 16; i++) {
+				if (!strcmp(colors[i], argv[2]))
+					sb_set_bg(sb + sb_index, i);
+			}
+		} else {
+			kprintf("color_change" ": Unknown command.\n");
+			return -1;
+		}
+	}
+	return 0;
```
Please provide a usage message instead of "not enough arguments", which is not very helpful. Also consider replacing "color_change" with the new function name: "color".

e2r3p13 commented on kernel/nsh/builtins/hexdump.c:
```diff
@@ -0,0 +1,105 @@
+
+int hexdump(int argc, char **argv) {
+	if (argc < 3) {
+		kprintf(BLTNAME ": Not enough argument.\n");
```
You can remove this message as you already print the usage

e2r3p13 commented on kernel/nsh/builtins/info.c:
```diff
@@ -0,0 +1,126 @@
+#define PRINT(reg, var, format) \
+	reg(var); kprintf(format, var)
+
+void info_registers()
+{
```
Please put opening braces on the same line that the function name to be consistent with other files. (same thing for all the functions in this file).

e2r3p13 commented on kernel/nsh/builtins/info.c:
```diff
@@ -0,0 +1,126 @@
+	} else if (!strcmp(argv[1], "registers")) {
+		info_registers();
+	} else {
+		kprintf("info : '%s' doesn't exist.\n", argv[1]);
+		kprintf("Usage : info [gdt/idt/stack/buddy/registers]\n");
```
Typo: There is no space before colons in english.

e2r3p13 commented on kernel/nsh/builtins/keymap.c:
```diff
@@ -0,0 +1,33 @@
+ */
+int keymap(int argc, char **argv) {
+	if (argc < 2) {
+		kprintf("keymap" ": Not enough arguments.\n");
```
Please replace this with a usage message

e2r3p13 commented on kernel/nsh/builtins/utils.c:
```diff
@@ -0,0 +1,64 @@
+/*
+ * Next screen.
+ */
+int next(__attribute__ ((unused)) int argc, __attribute__ ((unused)) char **argv) {
+	if (sb_index + 1 < sb_nbscreen) {
+		sb_unload(sb + sb_index);
+		sb_index++;
+		sb_load(sb + sb_index);
+		return 0;
+	}
+	return -1;
+}
+
+/*
+ * Next screen.
+ */
+int prev(__attribute__ ((unused)) int argc, __attribute__ ((unused)) char **argv) {
+	if (sb_index > 0) {
+		sb_unload(sb + sb_index);
+		sb_index--;
+		sb_load(sb + sb_index);
+		return 0;
+	}
+	return -1;
+}
+
```
Maybe we could display an error message if the screen switching is unsuccessful?
Or maybe we could loop over screens: If going prev while on screen 0 -> Go to screen `sb_nbscreen - 1`

xlmod replied:
I'm ok with both, but a loop seems a better option.

e2r3p13 commented on kernel/nsh/builtins/utils.c:
```diff
@@ -0,0 +1,64 @@
+ * Prints usage and help message.
+ */
+int help(__attribute__ ((unused)) int argc, __attribute__ ((unused)) char **argv) {
+	kprintf("KFS - available commands :\n");
```
No space before colons

e2r3p13 commented on kernel/nsh/builtins/utils.c:
```diff
@@ -0,0 +1,64 @@
+int help(__attribute__ ((unused)) int argc, __attribute__ ((unused)) char **argv) {
+	kprintf("KFS - available commands :\n");
+	for (int i = 0; builtin[i].name != NULL ; i++)
+		kprintf("-> %s\n", builtin[i].name);
```
Just an idea, but maybe we could store a tiny description of the builtin in the structure and display it here too?

xlmod replied:
Good idea, I already start thinking about this.

e2r3p13 commented on kernel/nsh/nsh.c:
```diff
@@ -0,0 +1,150 @@
+};
+
+
+static void nsh_prompt() {
+	uint8_t oldcolor;
+
+	oldcolor = sb_get_color(sb + sb_index);
+		sb_set_fg(sb + sb_index, SB_COLOR_LIGHT_BLUE);
+	kprintf("nsh$ ");
+	sb_set_color(sb + sb_index, oldcolor);
+}
```
Wow, this looks very nice! This guy codes very well...

xlmod replied:
mdr

e2r3p13 commented on kernel/nsh/nsh.c:
```diff
@@ -0,0 +1,150 @@
+	kprintf("nsh: %s: command not found\n", nsh_cmd[0]);
+}
+
+static void nsh_getkeyevent(struct kbd_event *evt) {
+	while (1) {
+		while (!KBD_poll());
+		if (KBD_getevent(evt) == 0 && evt->type == KEY_PRESSED)
+			return ;
+	}
+}
```
Shouldn't this function belongs to the keyboard module?

xlmod replied:
Perhaps add similar function but with the key event type in parameter.

e2r3p13 reviewed (changes requested):
Some coding style issues (Mostly for consistency). As a general comment:

- Opening braces at the end of the function name's line
- No spaces before colons in english (there is one in french)
- Comments are not always formatted in the same way in the code. We'll need to choose a style.
- We need to provide clear usage information for each builtins. Maybe even add a usage field in the structure and make use of it when the requirements are not met. We'll also need to standardize our usage format to make sure they're consistent.

The overall is pretty good, but note that I didn't read everything as carefully as it deserves. Another check would be nice, tho not essential.

xlmod committed:
Change request from pr #42

Add usage function in all builtin which took multiple arguments.
Remove all space before colon.
Move nsh_getkeyevent to keyboard module as KBD_getevent_by_type.
Add description for each builtin in builtin struct to print them in the
helper.

Signed-off-by: glafond- <glafond-@student.42.fr>

xlmod committed:
Add tabulation handler with autocompletion

Use the builtin struct array to compare nsh_buf to all builtin a fill it
with the first matching entry of the array.
If nsh_buf is empty do nothing.

Signed-off-by: glafond- <glafond-@student.42.fr>

mrxx0 commented on kernel/nsh/nsh.c:
```diff
@@ -0,0 +1,155 @@
+			kprintf("\b");
+		}
+	} else if (c == '\t') {
+		if (nsh_bufindex == 0)
```
Autocomplete like a boss ?? @lfalkau Take a look at that it's wild

mrxx0 reviewed (approved):
LGTM

mrxx0 commented on kernel/nsh/nsh.c:
```diff
@@ -0,0 +1,155 @@
+			nsh_buf[--nsh_bufindex] = 0;
+			kprintf("\b");
+		}
+	} else if (c == '\t') {
+		if (nsh_bufindex == 0)
+			return;
+		for (int i = 0; builtin[i].name != NULL; i++) {
+			char *p = nsh_buf;
+			char *q = builtin[i].name;
+			while (*p && *q && *p == *q)
+				p++, q++;
+			if (*p == 0 && *q != 0) {
+				kprintf("%s", q);
+				while (*q && nsh_bufindex < NSH_BUFSIZE)
+					nsh_buf[nsh_bufindex++] = *q++;
+			}
+		}
```
Autocomplete like a boss ?? @lfalkau Take a look at that it's wild

e2r3p13 commented on include/kernel/keyboard.h:
```diff
@@ -215,6 +215,7 @@ struct kbd_keymap_entry {
 void KBD_initialize();
 int KBD_poll();
 int KBD_getevent(struct kbd_event *evt);
+void KBD_getevent_by_type(struct kbd_event *evt, enum kbd_eventtype type);
```
I don't like the name of this function. Some words are split while some are not. For me it should either be `get_event_by_type` or `geteventbytype` (which I'd prefer). `geteventoftype` would also do the job`.

xlmod replied:
ok `geteventbytype` it's good to me.

e2r3p13 commented on kernel/nsh/nsh.c:
```diff
@@ -25,24 +25,23 @@ int nsh_cmdnarg;
+	{"exit", poweroff, "Alias of 'poweroff'"},
+	{"keymap", keymap, "Switch keyboard layout"},
+	{"color", color, "Change the screen buffer color"},
+	{"alloc", alloc, "Return a random address"},
```
Allocates a memory area of a random size (not random address)

e2r3p13 commented on kernel/nsh/nsh.c:
```diff
@@ -25,24 +25,23 @@ int nsh_cmdnarg;
+	{"keymap", keymap, "Switch keyboard layout"},
+	{"color", color, "Change the screen buffer color"},
+	{"alloc", alloc, "Return a random address"},
+	{"free", free, "Do nothing"},
```
It does something. It frees the last allocated block

mrxx0 committed:
Fix helper message, update alloc/free builtins...

- Update free and alloc helper description
- Update alloc builtins :
	- Takes a size as parameter
	- Allocate to fill the size with multiples chunks
- Update free builtins :
	- Takes an address and a size as parameter
- Fix kpm_alloc by checking the first order:
	- We were using a size_t type which skipped the first order
- Update KBD_getevent_by_type() function name to KBD_geteventbytype()

Signed-off-by: mrxx0 <chcoutur@student.42.fr>

mrxx0 committed:
Replace sb + sb_index with sb_current

Signed-off-by: mrxx0 <chcoutur@student.42.fr>

xlmod committed:
Change info buddy to print the right value

Signed-off-by: glafond- <glafond-@student.42.fr>

e2r3p13 commented on include/kernel/nsh.h:
```diff
@@ -0,0 +1,26 @@
+ * created: 2022/12/08 - xlmod <glafond-@student.42.fr>
+ * updated: 2022/12/14 - glafond- <glafond-@student.42.fr>
+ */
+
```
Two empty lines instead of one

e2r3p13 commented on include/kernel/screenbuf.h:
```diff
@@ -0,0 +1,77 @@
+#define SB_DFL_COLOR	(uint8_t)((SB_DLF_BG_COLOR << 4) | SB_DFL_FG_COLOR)// light grey on black
+#define SB_WHITESPACE	(uint16_t)(SB_DFL_COLOR << 8)
+
+#define SB_CURSOR_PORT1	0x3D4
+#define SB_CURSOR_PORT2	0x3D5
```
`SB_CURSOR_PORT1` -> `CURSOR_ADDR_PORT`
`SB_CURSOR_PORT2` -> `CURSOR_CMD_PORT`

e2r3p13 commented on kernel/gdt_internal.h:
```diff
@@ -6,12 +6,83 @@
+
+/* GDT_USER_STACK_ACCESS - 0xF6 */
+#define GDT_USER_STACK_ACCESS SEG_DESCTYPE(1) | SEG_PRES(1) | SEG_SAVL(0) | \
+                          SEG_LONG(0) | SEG_SIZE(1) | SEG_GRAN(1) | \
+                          SEG_PRIV(3) | SEG_DATA_RDWREXPD
 /* Describe an entry of the gdt.
```
Missing newline

e2r3p13 commented on kernel/nsh/builtins/alloc.c:
```diff
@@ -0,0 +1,67 @@
+	kprintf("Usage: " BLTNAME " size\n");
+}
+
+/* Wrapper for the kpm_alloc function.
+ * Loop until a maximum number of chunk has been allocated to allocate the 
+ * given memory size.
+ *
+ * */
```
Bad comment style

e2r3p13 commented on kernel/nsh/builtins/color.c:
```diff
@@ -0,0 +1,83 @@
+
+/* kernel/nsh/builtins/color.c
+ *
+ * Insert file description here
```
Insert file description

e2r3p13 commented on kernel/nsh/builtins/free.c:
```diff
@@ -0,0 +1,50 @@
+	kprintf("Usage: " BLTNAME " size\n");
+}
+
+/* Free a (previously allocated) memory area from addr on a given size in bytes.
+ * 
+ */
```
Bad comment style

e2r3p13 commented on kernel/nsh/builtins/hexdump.c:
```diff
@@ -0,0 +1,104 @@
+
+/* kernel/nsh/builtins/kexdump.c
+ *
+ * Insert file description here
```
File description

e2r3p13 commented on kernel/nsh/builtins/info.c:
```diff
@@ -0,0 +1,124 @@
+
+/* kernel/nsh/builtins/info.c
+ *
+ * Insert file description here
```
File description

e2r3p13 commented on kernel/nsh/builtins/interrupt.c:
```diff
@@ -0,0 +1,46 @@
+
+/* kernel/nsh/builtins/interrupt.c
+ *
+ * Insert file description here
```
File description

e2r3p13 commented on kernel/nsh/builtins/keymap.c:
```diff
@@ -0,0 +1,57 @@
+
+/* kernel/nsh/builtins/keymap.c
+ *
+ * Insert file description here
```
File desc

e2r3p13 commented on kernel/nsh/builtins/power.c:
```diff
@@ -0,0 +1,34 @@
+
+/* kernel/nsh/builtins/power.c
+ *
+ * Insert file description here
```
File desc

e2r3p13 commented on kernel/nsh/builtins/utils.c:
```diff
@@ -0,0 +1,61 @@
+
+/* kernel/nsh/builtins/utils.c
+ *
+ * Insert file description here
```
File desc

e2r3p13 commented on kernel/nsh/nsh.c:
```diff
@@ -0,0 +1,155 @@
+// SPDX-FileCopyrightText: CGL-KFS
+// SPDX-License-Identifier: BSD-3-Clause
+
+/* kernel/nsh/nsh.c
+ *
+ * Nulix shell
+ *
+ * created: 2022/12/07 - xlmod <glafond-@student.42.fr>
+ * updated: 2022/12/15 - mrxx0 <chcoutur@student.42.fr>
+ */
```
Please comment the whole file

e2r3p13 commented on kernel/screenbuf.c:
```diff
@@ -0,0 +1,349 @@
+/* kernel/screenbuf.c
+ *
+ * Big buffer to keep writing and add true scroll
```
Change description

e2r3p13 reviewed (changes requested):
Some typos and minor changes. I'll push force and merge right then.

xlmod committed:
Add comments in nsh.c and PAGE_UP/DOWN handle

Signed-off-by: glafond- <glafond-@student.42.fr>

e2r3p13 committed:
Fix typos and comments style

Signed-off-by: lfalkau <lfalkau@student.42.fr>
